frontend/src/templates/Field/Attachment/AttachmentField.tsx

```diff
@@ -70,6 +75,15 @@ export const AttachmentField = ({
         try {
           const buffer = await fileArrayBuffer(file)
           const clone = new File([buffer], file.name, { type: file.type })
+
+          /**
+           * Set a custom field to force attachment field to remain dirty.
+           * React Hook Form is unable to evaluate dirtiness file when comparing File objects https://react-hook-form.com/docs/useformstate#return
+           * React Hook Form has a custom deepEqual comparator function that checks for key values on the object https://github.com/react-hook-form/react-hook-form/blob/v7.51.1/src/utils/deepEqual.ts
+           * By introducing a new `__dirtyField` property, so we can set force the evaluation deepEqual to be false, thus remaining dirty.
+           * */
+          // @ts-expect-error __dirtyField is not a standard property of File.
+          clone.__dirtyField = 1
```

Comment:
Is this something mrf-specific or does this apply to all modes? It sounds like a generic statement

Also, I'm not sure how this helps in keeping the field dirty. What's the context for this?

Reply:
Update: I looked at the Notion doc again - just to be clear, so this is purely to support Singpass auth in the future? is it needed at this juncture?

Reply:
Specifically, I'm referring to this change [here](https://github.com/opengovsg/FormSG/pull/4275/files#diff-42654152f38821549821e8fd0cb30c57a7fadddc8a26b2bc714cd107c9c1f9b3R98).

1) It checks whether `formState.isDirty` and does a reset.
2) RHF is unable to evaluate dirtiness file when comparing `File` objects (read: when respondent changes the attachment field aka a `File` object, it will not be recognized as dirty by RHF)

Both 1 & 2 results in the form resetting to its defaultValues even when the respondent did change a file[a].

Then it begs the question, can we remove the offending code that does the reset? 
```
  useEffect(() => {
    if (!isDirty) {
      reset(defaultFormValues)
    }
  }, [defaultFormValues, isDirty, reset])
```

Well, no. Because it is meant to populate the form fields where the query params changes when user returns from the spcp login flow, thus this means that this code is performing as intended.

The question now becomes, what is not performing as intended. For this, I've deemed the `isDirty` to be incorrect, as it is reasonable for us to presume that `isDirty` should have been `true` here since the attachmentField has been updated, but had wrongfully identified that attachmentField to be not dirty.


[a] Walkthrough of how defaultValues + `isDirty` affects the attachment field.
```
1. Respondent 1 attaches a file, and submits.
2. Respondent 2 renders form
    - defaultValue['attachmentFieldId'] = someFileObj // as per [426541](https://github.com/opengovsg/FormSG/pull/7181/files#diff-42654152f38821549821e8fd0cb30c57a7fadddc8a26b2bc714cd107c9c1f9b3R111)
    - formValue['attachmentFieldId'] = someFileObj // as populated by defaultValue
    - formState['attachmentFieldId'].isDirty // false
4. Respondent 2 changes file 
    - defaultValue['attachmentFieldId'] = someFileObj
    - formValue['attachmentFieldId'] = newFileObj // as populated by defaultValue
    - formState['attachmentFieldId'].isDirty // false!! Because https://react-hook-form.com/docs/useformstate#return and the exact comparator is also linked above
```

Reply:
oh i see. the `reset` is necessary because of our query param handling within the spcp flow, since we restore the query params from browser storage in code. 🤦  

as a follow-up then, let's say R2 changes the file once, so that `__dirtyField = 1`. If R2 changes the file again, then `__dirtyField = 1` will continue to remain. will it work as intended in this case? it seems like it should still be dirty right? 

If so, would it make sense to do something like `__dirtyField = __dirtyField ? 0 : __dirtyField + 1` or something? (this would be capped by the browser maxint but like... no one is ever going to hit it)

or is my understanding still incorrect

Reply:
It will still be dirty, since rhf compares `formValue['attachmentFieldId'] ` with `defaultValue['attachmentFieldId']`. 

Thus no need to update `__dirtyField`.

Reply:
Ah. I gotcha. Interesting... I think this is fine haha just wanted to clarify the intention